go-controller/pkg/allocator/pod/macs.go

```diff
@@ -0,0 +1,157 @@
+	// Default to pod-based identifier
+	return fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+}
+
+// ReleasePodReservedMacAddress releases pod's reserved MAC address, if exists.
+// It removes the used MAC address, from pod network annotation, and remove it from the MAC manager store.
+func (allocator *PodAnnotationAllocator) ReleasePodReservedMacAddress(pod *corev1.Pod, mac net.HardwareAddr, claim *ipamclaimsapi.IPAMClaim) {
+	networkName := allocator.netInfo.GetNetworkName()
+	if allocator.macManager == nil {
+		klog.V(5).Infof("No mac manager for network %s, skipping mac address release", networkName)
+		return
+	}
+
+	if _, exist := pod.Labels[kubevirtv1.VirtualMachineNameLabel]; exist && claim != nil {
+		klog.V(5).Infof("Retaining MAC address (%s) of pod (%s/%s) on network (%s) becuase it has IPAM claim (%s/%s)",
+			mac, pod.Namespace, pod.Name, networkName, claim.Namespace, claim.Name)
+		return
+	}
+
+	owner := GetMACOwner(pod)
+	allocator.macManager.Release(owner, mac)
+
+	klog.V(5).Infof("Released MAC (%s) owned by (%s) on network (%s)", mac, owner, networkName)
+}
+
+// InitializeMACManager initializes MAC reservation tracker with MAC addresses in use in the network.
+func (allocator *PodAnnotationAllocator) InitializeMACManager() error {
+	networkName := allocator.netInfo.GetNetworkName()
+	if allocator.macManager == nil {
+		klog.V(5).Infof("No mac manager for network %s, skipping mac address release", networkName)
+		return nil
+	}
+
+	pods, err := allocator.fetchNetworkPods()
+	if err != nil {
+		return err
+	}
+	podMACs, err := getPodMACs(pods)
+	if err != nil {
+		return err
+	}
+
+	// reserve MCAs used by infra first, to prevent network disruptions to connected pods in case of conflict.
+	infraMACs := calculateSubnetsInfraMACAddresses(allocator.netInfo.Subnets())
+	for owner, mac := range infraMACs {
+		if err := allocator.macManager.Reserve(owner, mac); err != nil {
+			return fmt.Errorf("failed to reserve infra MAC (%s) for owner (%s) on network (%s): %w",
+				mac, owner, networkName, err)
+		}
+		klog.V(5).Infof("Reserved MAC (%s) for infra (%s) on network (%s)", mac, owner, networkName)
+	}
+	for owner, mac := range podMACs {
+		if rerr := allocator.macManager.Reserve(owner, mac); rerr != nil {
+			return fmt.Errorf("failed to reserve pod MAC (%s) for owner (%s) on network (%s): %w",
+				mac, owner, networkName, rerr)
+		}
+		klog.V(5).Infof("Reserved MAC (%s) for pod (%s) on network (%s)", mac, owner, networkName)
+	}
+
+	return nil
+}
+
+// calculateSubnetsInfraMACAddresses return map of the network infrastructure mac addresses and owner name.
+// It calculates the gateway (.2) and management (.1) ports MAC addresses from their IP address.
+func calculateSubnetsInfraMACAddresses(subnets []config.CIDRNetworkEntry) map[string]net.HardwareAddr {
+	reservedMACs := map[string]net.HardwareAddr{}
+	for _, subnet := range subnets {
+		if subnet.CIDR == nil {
+			continue
+		}
+
+		gwIP := util.GetNodeGatewayIfAddr(subnet.CIDR)
```

Comment:
you probably should use https://github.com/ovn-kubernetes/ovn-kubernetes/blob/a393d95c3eb0d9448ced69c5226836fa4b537c26/go-controller/pkg/util/multi_network.go#L793 instead. This takes into account the user's intents (e.g. the user actually configured the IP for the GW in the UDN).